hexf/main.swift

```diff
@@ -0,0 +1,188 @@
+//  hexf
+//
+//  Created as main.m by Kevin Wojniak on 9/24/17.
+//  Converted to main.swift by Reed Harston on 10/31/22.
+//  Copyright © 2017 ridiculous_fish. All rights reserved.
+//
+
+import Cocoa
+
+enum HexfError: Error {
+    case invalidUsage
+    case standardInputNoData
+    case launchAppNoUrl
+    case launchAppFailure
+}
+
+struct Controller {
+    private static let kAppIdentifier = "com.ridiculousfish.HexFiend"
+    
+    func printUsage() {
+        fputs("""
+Usage:
+
+  Open files:
+    hexf file1 [file2 file3 ...]
+
+  Compare files:
+    hexf -d file1 file2
+
+  Open piped data:
+    echo hello | hexf
+
+  Show help:
+    hexf -h | --help
+
+""", stderr)
+    }
+    
+    private static func standardize(path: String) -> String {
+        let url = URL(fileURLWithPath: path)
+        return url.path // get absolute path
+    }
+    
+    private var appRunning: Bool {
+        NSRunningApplication.runningApplications(withBundleIdentifier: Self.kAppIdentifier).first != nil
+    }
+    
+    private func launchApp(with args: [String]) throws {
+        guard let url = NSWorkspace.shared.urlForApplication(withBundleIdentifier: Self.kAppIdentifier) else {
+            fputs("Failed to get url to app bundle for: \(Self.kAppIdentifier)", stderr)
+            throw HexfError.launchAppNoUrl
+        }
+        
+        let config = [NSWorkspace.LaunchConfigurationKey.arguments: args]
+        
+        do {
+            // TODO: Heed deprecation warning, and get right config type. This will require an availability check.
+            //            try NSWorkspace.shared.openApplication(at: url, configuration: config)
+            try NSWorkspace.shared.launchApplication(at: url, options: NSWorkspace.LaunchOptions.default, configuration: config)
+        } catch {
+            fputs("Launch app failed: \(error.localizedDescription)", stderr)
+            throw HexfError.launchAppFailure
+        }
+    }
+    
+    private func processStandardInput() throws {
+        let inFile = FileHandle.standardInput
+        // TODO: Heed deprecation warning. This will require an availability check.
+        let data = inFile.readDataToEndOfFile()
+        
+        guard data.count != 0 else {
+            throw HexfError.standardInputNoData
+        }
+        
+        if self.appRunning {
+            // App is already running so post distributed notification
+            let center = DistributedNotificationCenter.default()
+            center.postNotificationName(NSNotification.Name(rawValue: "HFOpenDataNotification"),
+                                        object: nil,
+                                        userInfo: ["data" : data],
+                                        deliverImmediately: true)
+            return
+        }
+        
+        // App isn't running so launch it with custom args
+        try launchApp(with: [
+            "-HFOpenData",
+            data.base64EncodedString(options: .init(rawValue: 0))
```

Comment:
looks like this arg has a default value in swift, can we exclude it?
```suggestion
            data.base64EncodedString()
```

Reply:
I'm actually not sure. I'll have to test it.

From the comment in the code:
```swift
// I'm not sure what 0 is for the options, but that is what the Obj-C code passed in... so I kept it.
            //        NSString *base64Str = [data base64EncodedStringWithOptions:0];
```